src/audio/crossover/crossover.c

```diff
@@ -469,167 +461,85 @@ static int crossover_validate_config(struct comp_dev *dev,
+				uint32_t config_id, uint32_t *data_offset_size,
+				uint8_t *fragment, size_t fragment_size)
 {
-	struct comp_data *cd = comp_get_drvdata(dev);
-	int ret = 0;
+	struct comp_data *cd = module_get_private_data(mod);
+	struct sof_ipc_ctrl_data *cdata = (struct sof_ipc_ctrl_data *)fragment;
 
-	switch (cdata->cmd) {
-	case SOF_CTRL_CMD_BINARY:
-		comp_info(dev, "crossover_cmd_get_data(), SOF_CTRL_CMD_BINARY");
-		ret = comp_data_blob_get_cmd(cd->model_handler, cdata, max_size);
-		break;
-	default:
-		comp_err(dev, "crossover_cmd_get_data(), invalid command");
-		ret = -EINVAL;
-		break;
-	}
-	return ret;
-}
-
-/**
- * \brief Handles incoming IPC commands for Crossover component.
- */
-static int crossover_cmd(struct comp_dev *dev, int cmd, void *data,
-			 int max_data_size)
-{
-	struct sof_ipc_ctrl_data *cdata = ASSUME_ALIGNED(data, 4);
-	int ret = 0;
+	comp_info(mod->dev, "crossover_get_config()");
 
-	comp_info(dev, "crossover_cmd()");
-
-	switch (cmd) {
-	case COMP_CMD_SET_DATA:
-		ret = crossover_cmd_set_data(dev, cdata);
-		break;
-	case COMP_CMD_GET_DATA:
-		ret = crossover_cmd_get_data(dev, cdata, max_data_size);
-		break;
-	default:
-		comp_err(dev, "crossover_cmd(), invalid command");
-		ret = -EINVAL;
+	if (cdata->cmd != SOF_CTRL_CMD_BINARY) {
+		comp_err(mod->dev, "crossover_get_config(), invalid command");
+		return -EINVAL;
 	}
 
-	return ret;
-}
-
-/**
- * \brief Sets Crossover Filter component state.
- * \param[in,out] dev Crossover Filter base component device.
- * \param[in] cmd Command type.
- * \return Error code.
- */
-static int crossover_trigger(struct comp_dev *dev, int cmd)
-{
-	comp_info(dev, "crossover_trigger()");
-
-	return comp_set_state(dev, cmd);
+	return comp_data_blob_get_cmd(cd->model_handler, cdata, fragment_size);
 }
 
 /**
  * \brief Copies and processes stream data.
  * \param[in,out] dev Crossover Filter base component device.
  * \return Error code.
  */
-static int crossover_copy(struct comp_dev *dev)
+static int crossover_process_audio_stream(struct processing_module *mod,
+					  struct input_stream_buffer *input_buffers,
+					  int num_input_buffers,
+					  struct output_stream_buffer *output_buffers,
+					  int num_output_buffers)
 {
-	struct comp_data *cd = comp_get_drvdata(dev);
-	struct comp_buffer *source;
-	struct comp_buffer __sparse_cache *source_c;
 	struct comp_buffer *sinks[SOF_CROSSOVER_MAX_STREAMS] = { NULL };
-	struct comp_buffer __sparse_cache *sinks_c[SOF_CROSSOVER_MAX_STREAMS];
-	int i, ret = 0;
+	bool enabled_buffers[PLATFORM_MAX_STREAMS] = { false };
+	struct comp_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct audio_stream __sparse_cache *source = input_buffers[0].data;
 	uint32_t num_sinks;
 	uint32_t num_assigned_sinks = 0;
-	uint32_t frames = UINT_MAX;
-	uint32_t source_bytes, avail;
-	uint32_t sinks_bytes[SOF_CROSSOVER_MAX_STREAMS] = { 0 };
-
-	comp_dbg(dev, "crossover_copy()");
+	uint32_t frames = input_buffers[0].size;
```

Comment:
this assumes all buffers have the same space, won't we overrun?

Reply:
It should not happen. In this case case the function module_single_source_setup() checks the frames avail/free from all source and sink combinations: https://github.com/thesofproject/sof/blob/a38458dd2276fa24e37f5e60ef0a5613b03cbe00/src/audio/module_adapter/module_adapter.c#L778

Reply:
Definitely not intuitive, mind leaving a comment explaining that?

Reply:
@singalsu can you add a comment?

Reply:
Thanks, yep, the comment is now added!